doc/develguide.rst

```diff
@@ -0,0 +1,656 @@
+for reflecting Python exceptions back into Java appears.
+
+As all proxies are ties to Python references, this code is strongly tied to
+the reference queue.
+
+``JPClassLoader``
+++++++++++++++++++
+
+This code is responsible for loading the Java class thunks.  AS it is difficult
+to ensure we can access a Java jar from within Python, all Java native code
+is stored in a binary thunk compiled into the C++ layer as a header.  The 
+class loader provides a way to load this embedded jar first by bootstrapping
+a custom Java classloader and then using that classloader to load the internal
+jar.
+
+The classloader is mostly transparent.  It provides one method called findClass
+which loads a class from the internal jar. ::
+  
+  jclass JPClassLoader::findClass(string name)
+
+
+``JPEncoding``
++++++++++++++++
+
+Java concept of UTF is pretty much orthagonal to the rest of the world.  Java
+used 16 bits for its native characters.  But this was inadequate for all of the
+unicode characters, thus longer unicode character had to be encoded in the 
+16 bit space.  Rather the directly providing methods to convert to a standard
+encoding such as UTF8, Java used UTF16 encoded in 8 bits.  ``JPEncoding`` deals
+with converting this unusual encoding into something that Python can understand.
+
+The key method in this module is transcribe with signature ::
+
+  std::string transcribe(const char* in, size_t len,
+      const JPEncoding& sourceEncoding,
+      const JPEncoding& targetEncoding)
+
+There are two encodings provided, ``JPEncodingUTF8`` and ``JPEncodingJavaUTF8``.  
+By selecting the source and traget encoding transcribe can covert to or
+from Java to Python encoding.
+
+
+Java native code
+----------------
+
+At the lowest level of the onion is the native Java layer.  Although this 
+layer is most remote from Python, ironically it is the easiest layer to communicate
+with.  As the point of jpype is to communicate with Java, it is possible to 
+directly communicate with the jpype Java internals.  These can be imported
+from the package ``org.jpype``.  The code for the Java layer is located in
+``native/java``.  It is compiled into a jar in the build directory and then converted
+to a C++ header to be compiled into the ``_jpype`` module.
+
+The Java layer currently houses the reference queue, a classloader which can
+load a Java class from a bytestream source, the proxy code for implementing
+Java interfaces, and a memory compiler module which allows Python to directly 
+create a class from a string.
+
+  
+Tracing
+---------
+
+Because the relations between the layers can be daunting especially when things
+go wrong.  The CPython and C++ layer have a built in logger.  This logger 
+must be enabled with a compiler switch to activate.  To active the logger, touch 
+one of the cpp files in the native directory to mark the build as dirty, then
+compile the ``jpype`` module with: ::
+
+     python setup.py --enable-tracing devel
```

Comment:
I like this approach better than setting this particular environment variable

Reply:
The only tricky part is sometimes if tool is already built it will skip the source building entirely and ignore the switch.  Still working on that.